refactor: FrescoBasedTextInlineImageSpan

gouravkhunger opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

I've migrated `FrescoBasedTextInlineImageSpan.java` to kotlin. Reference #50513.

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Changelog:

[ANDROID] [CHANGED] - Refactor class `FrescoBasedTextInlineImageSpan` from Java to Kotlin.

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

## Test Plan:

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

Tested the RN tester app with `yarn android` on both new and old architecture.

mateoguzmana commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:
```diff
@@ -0,0 +1,160 @@
+    private val tintColor: Int,
+    uri: Uri?,
+    private val headers: ReadableMap?,
+    private val draweeControllerBuilder: AbstractDraweeControllerBuilder<*, *, ImageRequest, *>,
```
Try:

```suggestion
    private val draweeControllerBuilder: AbstractDraweeControllerBuilder<*, ImageRequest, *, *>,
```

Was a bit tricky to understand as well, but this worked

gouravkhunger replied:
Thank you for this! It worked. My bad I was passing ImageRequest as the third param while the class signature expects it to be on the 2nd:

https://github.com/facebook/fresco/blob/b79964050f2bd82c8e6e9bc5a96dcfe61f95ae67/drawee/src/main/java/com/facebook/drawee/controller/AbstractDraweeControllerBuilder.java#L36

mateoguzmana commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:
```diff
@@ -0,0 +1,161 @@
+import com.facebook.react.views.image.ImageResizeMode
+import com.facebook.react.views.text.internal.span.TextInlineImageSpan
+
+/**
+ * FrescoBasedTextInlineImageSpan is a span for Images that are inside <Text/>. It computes its size
+ * based on the input size. When it is time to draw, it will use the Fresco framework to get the
+ * right Drawable and let that draw.
+ *
+ * <p>Since Fresco needs to callback to the TextView that contains this, in the ViewManager, you
+ * must tell the Span about the TextView
+ *
+ * <p>Note: It borrows code from DynamicDrawableSpan and if that code updates how it computes size
+ * or draws, we need to update this as well.
+ */
```
Nitpick, <p> are not necessary for KDoc

```suggestion
/**
 * FrescoBasedTextInlineImageSpan is a span for Images that are inside <Text/>. It computes its size
 * based on the input size. When it is time to draw, it will use the Fresco framework to get the
 * right Drawable and let that draw.
 *
 * Since Fresco needs to callback to the TextView that contains this, in the ViewManager, you
 * must tell the Span about the TextView
 *
 * Note: It borrows code from DynamicDrawableSpan and if that code updates how it computes size
 * or draws, we need to update this as well.
 */
```

mateoguzmana commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:
```diff
@@ -0,0 +1,161 @@
+            draweeControllerBuilder.reset()
+
+            drawable = draweeHolder.topLevelDrawable!!
+            drawable!!.setBounds(0, 0, _width, _height)
+            drawable!!.takeIf { tintColor != 0 }?.setColorFilter(tintColor, PorterDuff.Mode.SRC_IN)
+            drawable!!.callback = this.textView
```
!! assumes the value will never be null — is that guaranteed here? If not, we might want to add a proper null check.

gouravkhunger replied:
Hi @mateoguzmana I've used that. PTAL

gouravkhunger replied:
Although we will have to use it here because kotlin won't infer that the object is asserted to be non null in the scope.

https://github.com/facebook/react-native/blob/e3136d22923bf7afbdca207d898e07a76a22b19b/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt#L166-L169

To avoid `!!` have to use a new variable like so:

```kotlin
val _drawable = checkNotNull(drawable) // might have to use val _drawable: Drawable
// _drawable can be used without !!
```

If you believe that's better I can do that. Thanks!

mateoguzmana replied:
I think that would be a good idea, let's go that way with the new variable 👍

mateoguzmana reviewed (changes requested):
Thanks for migrating this file! One small thing that would be nice to check about the null safety

mateoguzmana commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:
```diff
@@ -0,0 +1,172 @@
+            draweeHolder.controller = draweeController
+            draweeControllerBuilder.reset()
+
+            checkNotNull(draweeHolder.topLevelDrawable).let { _drawable ->
+                _drawable.setBounds(0, 0, _width, _height)
+                if (tintColor != 0) {
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                        _drawable.colorFilter = BlendModeColorFilter(tintColor, BlendMode.SRC_IN)
+                    } else {
+                        _drawable.setColorFilter(tintColor, PorterDuff.Mode.SRC_IN)
+                    }
+                }
+                _drawable.callback = textView
+                drawable = _drawable
+            }
```
Looks way cleaner now! if you want to improve it even further, you can avoid using `_drawable` as `apply` scopes the variable into the block. Something like this:

```suggestion
          checkNotNull(draweeHolder.topLevelDrawable).apply {
            setBounds(0, 0, _width, _height)
            if (tintColor != 0) {
                colorFilter = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                    BlendModeColorFilter(tintColor, BlendMode.SRC_IN)
                } else {
                   setColorFilter(tintColor, PorterDuff.Mode.SRC_IN)
                }
            }
            callback = textView
            drawable = this
        }
```

gouravkhunger replied:
I had tried that implementation haha. Should've kept it:

https://github.com/facebook/react-native/blob/16d992c6140bacf90db9dfdef4cf7afa7f3ffa92/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt#L145-L156

let me revert to that.

gouravkhunger replied:
done

gouravkhunger commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:
```diff
@@ -0,0 +1,173 @@
+ * Note: It borrows code from DynamicDrawableSpan and if that code updates how it computes size or
+ * draws, we need to update this as well.
+ */
+public class FrescoBasedReactTextInlineImageSpan(
```
@mateoguzmana The CI build failed. Maybe the class should've been private?

```
> Task :packages:react-native:ReactAndroid:configureCMakeRelWithDebInfo[arm64-v8a]
e: file:///__w/react-native/react-native/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:53:5 Subclass 'public' exposes its 'internal' supertype 'TextInlineImageSpan'.
```

mateoguzmana replied:
Was just checking this, the direct conversion from Java would be to make this class internal:

```suggestion
internal class FrescoBasedReactTextInlineImageSpan(
```

gouravkhunger replied:
Alright, thanks!

gouravkhunger commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:
```diff
@@ -0,0 +1,173 @@
+            checkNotNull(draweeHolder.topLevelDrawable).apply {
+                setBounds(0, 0, _width, _height)
+                if (tintColor != 0) {
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                        colorFilter = BlendModeColorFilter(tintColor, BlendMode.SRC_IN)
+                    } else {
+                        setColorFilter(tintColor, PorterDuff.Mode.SRC_IN)
+                    }
```
@mateoguzmana I had added that sdk version check to account for the function deprecation. In my experience in other projects, usually the warning resolves with that check. But still getting:

```
> Task :packages:react-native:ReactAndroid:configureCMakeRelWithDebInfo[arm64-v8a]
e: warnings found and -Werror specified
w: file:///__w/react-native/react-native/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageSpan.kt:151:25 'fun setColorFilter(p0: Int, p1: PorterDuff.Mode): Unit' is deprecated. Deprecated in Java.
```

Apart from that warning app did compile and run fine locally.

mateoguzmana replied:
You can try:

```suggestion
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                        colorFilter = BlendModeColorFilter(tintColor, BlendMode.SRC_IN)
                    } else {
                        colorFilter = PorterDuffColorFilter(tintColor, PorterDuff.Mode.SRC_IN)
                    }
```

In case we need to keep it as-is, you can add on top of the line `@Suppress("DEPRECATION")` for now.

gouravkhunger replied:
I dont get the issue locally now. I think should be fine on CI too 👍

gouravkhunger commented:
Hi @mateoguzmana @mdvacca gentle nudge. Is there anything else you need from my side? Thank you so much!

mateoguzmana commented:
@gouravkhunger, nothing else is needed from your side for now. Once the second reviewer has availability, this will be imported and merged if there are no further things to fix

cortinico reviewed (changes requested):
@gouravkhunger can you rebase and fix the merge conflicts/build failures you'll have?

mateoguzmana commented:
> Hi @cortinico I've rebased but I'm now facing some type issues incompatibility of `AbstractDraweeControllerBuilder` between this file and [`FrescoBasedReactTextInlineImageShadowNode.kt`](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageShadowNode.kt). Things were alright before the linked file was migrated to Kotlin.
> 
> I tried fixing the type in that file- and changing it in mine to make both the same but i keep getting a variation of the error
> 
> ```
> BUILD FAILED in 5s
> error Failed to install the app. Command failed with exit code 1: ./gradlew installHermesDebug -PreactNativeDevServerPort=8081
> e: file:///Users/gourav/temp/react-native/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/frescosupport/FrescoBasedReactTextInlineImageShadowNode.kt:119:9 Argument type mismatch: actual type is 'com.facebook.drawee.controller.AbstractDraweeControllerBuilder<CapturedType(*)#4, CapturedType(*)#3, CapturedType(*)#2, CapturedType(*)#1>', but 'com.facebook.drawee.controller.AbstractDraweeControllerBuilder<*, com.facebook.imagepipeline.request.ImageRequest, *, *>' was expected. FAILURE: Build failed with an exception. * What went wrong:
> Execution failed for task ':packages:react-native:ReactAndroid:compileDebugKotlin'.
> ```
> 
> Could you please help point to the right direction? Thank you so much!

It's possible that the Kotlin compiler now expects more precise or explicit types than the original Java code (that file you are having problems with was just migrated to Kotlin as well).

Can you try updating the types for `draweeControllerBuilder` in FrescoBasedReactTextInlineImageShadowNode and FrescoBasedReactTextInlineImageViewManager? That might fix your problem

gouravkhunger commented:
Thank you @mateoguzmana! I could try in a few hours. Will revert back as soon as possible 👍

gouravkhunger commented:
@mateoguzmana The isssue indeed lied in `FrescoBasedReactTextInlineImageViewManager`. I hadn't come across it. As soon as I updated the types it went through alright 🎉 

@cortinico Please take a look and lmk if there's something else needed.

P.S. Since I rebased I've been getting some CMake warnings regarding

```
C/C++: CMake Warning (dev) at <path>/react-native/packages/react-native/ReactAndroid/build/third-party-ndk/googletest/googletest/cmake/internal_utils.cmake:152 (add_library):
Policy CMP0069 not being set
...
This warning is for project developers.  Use -Wno-dev to suppress it.
....
```

I hope this doesn't affect the code in this PR?

cortinico commented:
> P.S. Since I rebased I've been getting some CMake warnings regarding

Yeah I've looked into those and should be fixed by now
